### PR TITLE
Do not call grpc_init() for per-call-completion-queues created by a C++ sync server

### PR DESCRIPTION
**Pull request by sreecha**
Issue: #9261

Before this PR, every C++ `CompletionQueue` object creation ended up calling [`grpc_init`](https://github.com/grpc/grpc/blob/master/src/core/lib/surface/init.c#L176)  which acquires a mutex `g_init_mu` to increment the number of grpc-init invocations.

This is ok if the number of completion queues being created was small. However, in case of C++ Synchronous server, a completion queue is create for *every call*. This ends up making `g_init_mu` one of top created mutexes in sync servers.

This PR makes calling `grpc_init` optional when creating C++ `CompletionQueue` objects. It is always set to `true` by default but if a C++ `CompletionQueue`  is being created by passing a `grpc_completion_queue *` pointer i.e the following constructor (which is what Sync C++ server calls when creating a per-call completion queue), it is okay to NOT call `grpc_init`.

``` C++
class CompletionQueue : private GrpcLibraryCodegen {
 ..
  explicit CompletionQueue(grpc_completion_queue* take);
..
}
```

This change is safe because if we got a `grpc_completion_queue *` structure, we MUST have already called `grpc_init()`.

__Benchmark results__
I added a new benchmark BM_CreateDestroyCpp2 (didn't previously include it as a part of the PR because I thought it is trivial - but on a second thought, i think it is a good idea :))

Locks are down from 3 to 1

On master (with BM_CreateDestroyCpp2 temporarily ported):
``` 
sreek@sreek-dev:~/workspace/grpc (master) $ bins/counters/bm_cq --benchmark_filter=BM_CreateDestroyCpp2
Run on (12 X 3501 MHz CPU s)
2017-03-29 10:54:34
Benchmark                     Time           CPU Iterations
-----------------------------------------------------------
BM_CreateDestroyCpp2        215 ns        215 ns    3230778 locks/iter:3 atm_cas/iter:0 atm_add/iter:3 allocs/iter:1
```

On this PR branch:
```
sreek@sreek-dev:~/workspace/grpc (init-free-cq) $ bins/counters/bm_cq --benchmark_filter=BM_CreateDestroyCpp2
Run on (12 X 3501 MHz CPU s)
2017-03-29 10:56:15
Benchmark                     Time           CPU Iterations
-----------------------------------------------------------
BM_CreateDestroyCpp2        160 ns        160 ns    4413813 locks/iter:1 atm_cas/iter:0 atm_add/iter:3 allocs/iter:1
```

### CODE REVIEW

File: include/grpc++/impl/codegen/grpc_library.h [hunk 1]
```diff
@@ -51,18 +51,26 @@ extern GrpcLibraryInterface* g_glip;
...
+    if (grpc_init_called) {
+      GPR_CODEGEN_ASSERT(g_glip &&
+                         "gRPC library not initialized. See "
+                         "grpc::internal::GrpcLibraryInitializer.");
+      g_glip->shutdown();
+    }
   }
+
+ private:
+  bool grpc_init_called;
```
Comment (vjpai) [inline include/grpc++/impl/codegen/grpc_library.h]: Member variable should end with underscore

**Comment by vjpai** [comment]
Beside the minor comment above, this seems LGTM-able. Can you put benchmark results about this into this PR description?

**Comment by sreecha** [comment]
@vjpai, Did the changes you requested. PTAL

**Comment by sreecha** [comment]
`gRPC_interop_pull_requests` failure is a known issue: https://github.com/grpc/grpc/issues/9011
(`Bazel Full Build & Tests` suite is not fully implemented yet and we are ignoring those failures)

**Pull request merged**